### ISSUE TITLE
[Bug] Wrong highlight for shell

### ISSUE THREAD

**Issue by ricenoddle** [microsoft/monaco-editor#2899]
### Reproducible in vscode.dev or in VS Code Desktop?

- [X] Not reproducible in [vscode.dev](https://vscode.dev) or VS Code Desktop

### Reproducible in the monaco editor playground?

- [ ] Not reproducible in [the monaco editor playground](https://microsoft.github.io/monaco-editor/playground.html)

### Monaco Editor Playground Code

```typescript
Do not use playground to try this. Using home page editor demo to try the following code.


if [[ -z "${MA5800_ALL_DIR:-}" ]]; then
    MA5800_ALL_DIR=$(realpath "$(dirname "$0")/../../")
fi
```

### Actual Behavior

![image](https://user-images.githubusercontent.com/13479751/150057404-5bc3c6c4-85f5-47e6-840f-62d07ce1a11e.png)

So I tried this code in vs-code, the highlight is correct.

![image](https://user-images.githubusercontent.com/13479751/150057771-7ec92c9b-9dd8-464f-9c3e-4d6fd81add8d.png)

### Expected Behavior

My expected behavior is the highlight can as same as vs-code

**Comment by hediet** [comment]
Looks like the last quotation mark is also wrong.
We're more than happy to accept a PR for this!

**Comment by ricenoddle** [comment]
Yes, looks like the last quote make the `monarch` think the quotation is not finished

**Comment by Q849647178** [comment]
I also encountered this problem yesterday and solved it by using setMonarchTokensProvider. I'll keep a record of it.
The general code is as follows
`
monaco.languages.register({ id: 'customShell' });
      monaco.languages.setMonarchTokensProvider('customShell', shellMonarch);
      monaco.editor.defineTheme('shellTheme', {
        base: 'vs-dark',
        inherit: true,
        rules: [
          { token: 'variable', foreground: '0077b6' },
          { token: 'constant', foreground: '2a9d8f' },
          { token: 'keyword', foreground: 'c77dff' },
          { token: 'identifier', foreground: 'd6ce93' },
          { token: 'operator', foreground: '669bbc' },
        ],
        colors: {},
      });
      theme = 'shellTheme';
`

`export const shellMonarch = {
  tokenizer: {
    root: [
      [/#.*$/, 'comment'],

      [/"/, 'string', '@string'],
      [/'([^'\\]|\\.)*'/, 'string'],

      [/\b(true|false)\b/, 'constant.boolean'],
      [/\b[0-9]+\b/, 'constant.numeric'],
      [/\b[A-Z_][A-Z0-9_]*\b/, 'constant'],

      [/\${?[a-zA-Z_][a-zA-Z0-9_]*}?/, 'variable'],

      [/(if|then|else|elif|fi|for|while|do|done|case|esac|function|return|exit|echo)\b/, 'keyword'],

      [/[a-zA-Z_][a-zA-Z0-9_]*/, 'identifier'],

      [/[=<>!]+/, 'operator'],

      { include: '@whitespace' },
    ],

    string: [
      [/"/, 'string', '@popall'],
      [/\${?[a-zA-Z_][a-zA-Z0-9_]*}?/, 'variable'],
      [/[^\\$"]+/, 'string'],
      [/\\./, 'string.escape'],
    ],

    whitespace: [
      [/[ \t\r\n]+/, ''],
      [/\\[\r\n]+/, ''],
    ],
  },
};
`
